DESIGN.md

```diff
@@ -0,0 +1,193 @@
+# Technical Design
+
+This document outlines the technical background and design decisions for tinygraph, a library optimized for compact and efficient graphs using succinct data structures.
+
+## Introduction
+
+tinygraph explores the design space of sparse graphs, such as road networks, with a focus on making graphs both compact and efficient for modern hardware.
+As computation becomes cheaper, memory remains a significant bottleneck.
+tinygraph aims to optimize graphs to fit into random access memory or even the L3 cache using succinct data structure and compression techniques.
+
+This document provides a top-down introduction to graph representations, the design decisions behind tinygraph, and the optimizations used to achieve compactness and efficiency.
+
+
+## Graph Representations
+
+We use the following graph as a working example below:
+
+```mermaid
+flowchart TD
+  0@{ shape: circle, label: "0" }
+  1@{ shape: circle, label: "1" }
+  2@{ shape: circle, label: "2" }
+
+  0 --> 1
+  0 --> 2
+  1 --> 0
+  1 --> 2
+  2 --> 1
+```
+
+### Adjacency List
+
+An adjacency list represents a graph as a collection of linked lists, where each node points to its neighbors.
+
+**Example:**
+
+```
+0 -> [1, 2]
+1 -> [0, 2]
+2 -> [1]
+```
+
+Or in Python
+
+```python
+g = [list() for _ in range(3)]
+
+g[0].append(1)
+g[0].append(2)
+g[1].append(0)
+g[1].append(2)
+g[2].append(1)
+```
+
+**Pros:**
+1. Simple to implement
+2. Good enough for small graphs
+
+**Cons:**
+1. High space overhead due to pointer-based lists
+2. Poor memory locality, leading to cache misses during traversal
+3. Inefficient for large graphs due to random memory accesses
+
+
+### Adjacency Matrix
+
+An adjacency matrix represents a graph as a binary matrix where `matrix[i][j]` indicates an edge between nodes `i` and `j`.
+
+**Example:**
+
+```
+[
+  [0, 1, 1],
+  [1, 0, 1],
+  [0, 1, 0]
+]
+```
+
+Or in Python:
+
+```python
+g = [[0 for _ in range(3)] for _ in range(3)]
+
+g[0][1] = 1
+g[0][2] = 1
+g[1][0] = 1
+g[1][2] = 1
+g[2][1] = 1
+```
+
+**Pros:**
+- Simple and cache-efficient for dense graphs
+- Fast edge lookup and traversal
+
+**Cons:**
+- Quadratic space complexity, which is impractical for large sparse graphs
+- Wasteful for sparse graphs, as most entries are zeros
+
+
+### Compressed Sparse Row (CSR)
+
+CSR is a memory-efficient representation for sparse graphs.
+
+It stores edges in two arrays:
+1. `targets`: The target nodes of edges
+2. `offsets`: The starting and ending indices of edges for each source node
+
+**Example:**
+
+```
+targets = [1, 2, 0, 2, 1]
+offsets = [0, 2, 4, 5]
```

Comment:
Isn't this also a case where you can use Elias-Fano representation? Some quick calculations:

* Lets say we have `n = 100_000` nodes where each has 5 neighbors.
* This means that that `targets` end up being `u = 500_000`. This will be the largest value in `offsets`.
* With Elias-Fano you can encode `offsets` use `~log2(u / n)` bits for the lower part. In our example I believe 2 bits is optimal.
* This leads to us using `2 * n` bits for the lower bits and `(u >> 2) + n` bits for  the higher parts.
* This ends up taking **425000 bits**.

In comparison, The Succinct Compressed Sparse Row would need **500000 bits** since it needs one bit per neighbor.